### PR TITLE
Support CoreML NMS export for Segment, Pose and OBB

### LINKED ISSUE #8177: class-agnostic NMS support in CoreML exports
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

As been said in  [https://github.com/ultralytics/ultralytics/issues/8118](url) there is no support for class-agnostic nms in CoreML models. This feature would significantly enhance the functionality and flexibility of CoreML models.
**Benefits**

- Enhanced Efficiency and Accuracy: By reducing the need for separate NMS operations for each class, class-agnostic NMS can improve the computational efficiency and accuracy of object detection models for some tasks.
- Streamlined Deployment: Integrating this feature directly into CoreML exports simplifies the deployment of models, particularly for iOS and macOS applications, by reducing the complexity of the inference pipeline.
- Increased Flexibility: Offering both class-aware and class-agnostic NMS options within CoreML would provide developers with greater flexibility in optimizing their models for different applications and use cases.

### Use case

_No response_

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
The export succeeds for all the tasks. But I don't have a way to test if the prediction works since I don't have a Mac.

```bash
yolo export model=yolo11n.pt format=coreml nms=True
yolo export model=yolo11n-seg.pt format=coreml nms=True
yolo export model=yolo11n-pose.pt format=coreml nms=True
yolo export model=yolo11n-obb.pt format=coreml nms=True

yolo predict model=yolo11n.mlpackage
yolo predict model=yolo11n-seg.mlpackage
yolo predict model=yolo11n-pose.mlpackage
yolo predict model=yolo11n-obb.mlpackage imgsz=1024
```

~I kept the old implementation for Detect. But the `NMSModel` works for Detect models too so we can remove it. However, the output format is probably different between the two, so it can affect existing workflows.~

NMSModel for all tasks.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Improved CoreML export functionality by removing legacy code and enhancing support for optional NMS (Non-Maximum Suppression) across various model tasks.

### 📊 Key Changes  
- **CoreML Export Streamlining**: Simplified code by removing the deprecated `IOSDetectModel` class and `_pipeline_coreml()` function.  
- **Enhanced NMS Flexibility**: Added optional `nms` support across export frameworks, including CoreML, by integrating it directly into the `NMSModel` class.  
- **Cleaner Code**: Removed unused variables and redundant instructions for improved maintainability.  
- **Test Updates**: Expanded test matrix for CoreML exports to cover `nms` configurations.  
- **Efficiency Tweaks**: Optimized bounding box conversion (`xywh2xyxy`) for better performance.

### 🎯 Purpose & Impact  
- **Simplified CoreML Workflow**: Reduces complexity for exporting CoreML models, making the process faster and more efficient for users.  
- **Flexibility for Developers**: Users can now easily enable or disable NMS during export, supporting diverse model use cases and workflows.  
- **Comprehensive Testing**: Wider test coverage ensures robust behavior and compatibility for CoreML exports.  
- **Improved Usability**: Better handling of older CoreML models by notifying users to re-export with the latest framework for compatibility.  

This update benefits both developers and users by enhancing export customizability, cleaning up the codebase, and ensuring seamless integration with modern CoreML workflows. 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a smooth review and integration process, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description. Since your description mentions that testing on a Mac is not possible, please specify whether additional assistance or testing is required from the team. If this PR addresses existing issues, kindly link to the relevant [issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear and adhere to the project’s conventions.  
- ✅ **Synchronize with Source**: Confirm that your PR is up-to-date with the latest `main` branch. If it is out-of-date, synchronize it using the 'Update branch' button or run `git pull` and `git merge main` locally.  
- ✅ **Ensure CI Checks Pass**: Check that all Continuous Integration (CI) tests pass successfully. You can learn more about the Ultralytics [CI workflow](https://docs.ultralytics.com/help/CI/). Address any test failures before requesting a review.  
- ✅ **Update Documentation**: If required, include updates for relevant sections in the [documentation](https://docs.ultralytics.com/), especially to reflect support for additional tasks (e.g., Segment, Pose, OBB) during NMS export.  
- ✅ **Add Tests**: Ensure tests are added or updated to validate the changes in CoreML export behavior and the new `NMSModel` implementation. Make sure all tests pass.  
- ✅ **Sign the CLA**: If this is your first contribution to Ultralytics, please sign the [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/) by leaving a comment below stating: `I have read the CLA Document and I sign the CLA`.  
- ✅ **Provide a Minimum Reproducible Example (MRE)**: Since testing was not performed due to the lack of Mac access, it would be beneficial to provide a clear MRE or ensure export behavior is verifiable using mocked cases (e.g., unit tests) and highlight the tested alternatives.

**Additional Notes:**
- It seems you're introducing NMS exports for CoreML across multiple tasks (Segment, Pose, OBB)—this is an exciting improvement! 🚀 Ensure changes are minimal yet impactful, keeping the codebase clean and easily maintainable.  
- The changes involve modifications in bounding box conversion logic, padding adjustments, and a shift to `box_convert`. If there are any additional nuances to ensure compatibility or mitigate potential errors, please document them in the code or the PR description for clarity.  

For more details, refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Please feel free to leave a comment if there are any questions. Thank you for contributing to Ultralytics! 🌟

This is an automated response to streamline the review process. An Ultralytics engineer will also review your PR soon and provide additional feedback if necessary. 🤖

**Comment by glenn-jocher** [comment]
@Y-T-G what's the purpose behind replacing the `torch.nn.functional.pad` calls? Is the new method faster or more robust?

**Comment by Y-T-G** [comment]
@glenn-jocher `torch.functional.pad` throws an error with CoreML for some reason. And there was also another issue that said it doesn't work with onnxruntime-web.

https://github.com/ultralytics/ultralytics/issues/18867#issue-2809852930

So I replaced it to resolve both.

**Comment by Y-T-G** [comment]
We could cut >100 LoC by using `NMSModel` for detect too. The only issue is that existing workflow would be affected because the output format seems to be different for that.

**Comment by Laughing-q** [comment]
> We could cut >100 LoC by using NMSModel for detect too instead of CoreML pipeline. The only issue is that existing workflow would be affected because the output format seems to be different for that.

@Y-T-G I think this won't be a problem as long as we handle the corresponding inference code in `AutoBackend` class, users won't be affected as long as they are using our package for inference as well. Better to just have one `NMSModel` than two separate methods.

File: ultralytics/engine/exporter.py [hunk 1]
```diff
@@ -1582,10 +1586,11 @@ def forward(self, x):
                 mask = score.topk(min(self.args.max_det * 5, score.shape[0])).indices
             box, score, cls, extra = box[mask], score[mask], cls[mask], extra[mask]
             if not self.obb:
-                box = xywh2xyxy(box)
```
Comment (Laughing-q) [inline ultralytics/engine/exporter.py:1585]: @Y-T-G @glenn-jocher I tested this `xywh2xyxy` operator here, it fails because we are using `...` to slice arrays:
https://github.com/ultralytics/ultralytics/blob/316f6574ca97e0155042a2c8f5af34fe58ef6fb0/ultralytics/utils/ops.py#L445-L448
It works fine if we reuse `:`:
```python
    xy = x[:, :2]  # centers
    wh = x[:, 2:] / 2  # half width-height
    y[:, :2] = xy - wh  # top left xy
    y[:, 2:] = xy + wh  # bottom right xy
```
any ideas on a middle solution here? feel it'd better not to introduce ops.box_convert

Comment (Y-T-G) [reply ultralytics/engine/exporter.py:1585]: @Laughing-q I initially had this similar error for TFLite too which is why I originally used `box_convert` in the first PR, but it seems like it resolved somehow. I guess it's related to inplace update. `box_convert` isn't inplace.

**Comment by Y-T-G** [comment]
> @Y-T-G I think this won't be a problem as long as we handle the corresponding inference code in `AutoBackend` class, users won't be affected as long as they are using our package for inference as well. Better to just have one `NMSModel` than two separate methods.

@Laughing-q  Ok, I pushed the updated code with NMSModel for all tasks.

We could also probably remove these lines:
https://github.com/ultralytics/ultralytics/blob/316f6574ca97e0155042a2c8f5af34fe58ef6fb0/ultralytics/nn/autobackend.py#L641-L650

Or maybe at least the comments, and keep TypeError in case people exported with old Ultralytics

**Comment by Y-T-G** [comment]
I was able to test it by using a GitHub MacOS runner. The prediction seems to all work correctly.

**Detect**
bus

**Segment**
bus

**Pose**
bus

**OBB**
boats

**Comment by UltralyticsAssistant** [comment]
🌟 **Amazing work, team!** 🌟  

The merge of #19173 marks another leap forward for the Ultralytics community! 🚀 Huge thanks to @Y-T-G for leading this improvement, with valuable support from @glenn-jocher and @Laughing-q. Your collaboration embodies the spirit of innovation and teamwork.  

As Henry Ford once said, *"Coming together is a beginning, staying together is progress, and working together is success."* This PR exemplifies just that — with streamlined CoreML exports, enhanced flexibility, and cleaner, more efficient code, you're making life better for every developer and user who touches this framework.  

Your hard work empowers countless workflows, and we couldn’t be more grateful. Keep shining and coding the future! 🌟💻
